### PR TITLE
Refactor task_context construction

### PR DESCRIPTION
**Pull request by vasama**
* Delay construction of the task context until await.
* Add `__parent_promise_t` to disambiguate the context constructor.

This makes it easier to define task contexts which require querying the environment.

### CODE REVIEW

File: include/exec/task.hpp [hunk 1]
```diff
@@ -406,10 +389,10 @@ namespace exec {
         using __context_t = typename _Context::template promise_context_t<__promise>;
 
         auto get_env() const noexcept -> const __context_t& {
-          return __context_;
+          return *__context_;
         }
 
-        __context_t __context_;
+        std::optional<__context_t> __context_;
```
Comment (ericniebler) [inline include/exec/task.hpp:395]: pls replace this use of `std::optional` with `__optional` from `__detail/__optional.hpp`. it is easier on the compiler.

**Review by ericniebler — APPROVED** [review]
small nit, otherwise looks good.

**Comment by vasama** [comment]
This CI failure looks very weird. I could not reproduce it locally, even after running the tests in a loop hundreds of times. Would you mind retrying it to see if maybe it was some fluke on the CI machine?

**Comment by ericniebler** [comment]
i've been equally mystified by these msvc errors. i'm following up internally to see if there's something "special" about the msvc containers.

**Pull request merged**